ntp-client/src/Network/NTP/Client.hs

```diff
@@ -71,248 +64,221 @@ data NtpStatus =
-                case sock of
-                    This (Last sock_)
-                        -> loop addressFamily $ runWithAddrFamily sock_
-                    That (Last sock_)
-                        -> loop addressFamily $ runWithAddrFamily sock_
-                    These _ _
-                        -> error "NtpClient: startReceive: impossible"
-
-    -- Compute the clock offset based on current time and record it in the NTP
-    -- client state.   A packet will be disgarded if it came after
-    -- @'ntpResponseTimeout'@.
-    handleNtpPacket
-        :: NtpPacket
-        -> IO ()
-    handleNtpPacket packet = do
-        traceWith tracer NtpTraceReceiveLoopPacketReceived -- packet
-        clockOffset (ntpResponseTimeout $ ncSettings cli) packet >>= \case
-            Nothing -> traceWith tracer NtpTraceReceiveLoopLatePacket
-            Just offset -> do
-                traceWith tracer $ NtpTraceReceiveLoopPacketDeltaTime $ getNtpOffset offset
-                atomically $ modifyTVar' (ncState cli) ( offset : )
-
--- |
--- Spawn NTP client which will send request to NTP servers every @'ntpPollDelay'@
--- and will listen for responses.  The @'ncStatus'@ will be updated every
--- @'ntpPollDelay'@ with the most recent value.  It should be run in a separate
--- thread, since it will block infinitely.
-spawnNtpClient :: Tracer IO NtpTrace -> NtpClientSettings -> TVar NtpStatus -> IO ()
-spawnNtpClient tracer settings ncStatus = do
-    traceWith tracer NtpTraceSpawnNtpClientStarting
-    bracket (mkSockets tracer settings) closeSockets $ \sock -> do
-        cli <- mkNtpClient settings ncStatus sock
-
-        addrs <- resolve
-        -- TODO
-        -- we should start listening for requests when we send something, since
-        -- we're not expecting anything to come unless we send something.  This
-        -- way we could simplify the client and remove `ncState` mutable cell.
-        receiveLoop tracer cli
-            `concurrently_` sendLoop tracer cli addrs
-            `concurrently_` traceWith tracer NtpTraceSpawnNtpClientStarted
-
+        restartDelay :: [Int]
+        restartDelay = [0, 5, 10, 20, 60, 180, 600] ++ repeat 600
+
+-- | Setup and run the NTP client.
+-- In case of an IOError (for example when network interface goes down) cleanup and return.
+
+oneshotClient ::
+       Tracer IO NtpTrace
+    -> (NtpClientSettings, TVar NtpStatus)
+    -> IO ()
+oneshotClient tracer (ntpSettings, ntpStatus) = do
+    traceWith tracer NtpTraceClientStartQuery
+    (v4Servers,   v6Servers)   <- lookupServers $ ntpServers ntpSettings
+    (v4LocalAddr, v6LocalAddr) <- udpLocalAddresses >>= firstAddr "localhost"
+    (v4Replies, v6Replies) <- concurrently (runProtocol IPv4 v4LocalAddr v4Servers)
+                                           (runProtocol IPv6 v6LocalAddr v6Servers)
+    when (null v4Replies && null v6Replies) $ do
+        traceWith tracer NtpTraceIPv4IPv6BothFailed
+        ioError $ userError "IPv4 and IPv6 failed"
```

Comment:
I want to make the case where there are no replies from IPv4 nor IPv6 an error that
propagates one level up and causes the client to be respawn (after 10 s) instead of just a regular query.
In the other case there was at least one reply and (maybe) its is more OK to just wait
for 30 minutes and try again.

Reply:
If there are no replies that should be visible without throwing exceptions.  Just pattern matching on the list of results.

Reply:
OK I changed it:
new behavior: No difference between no replies and errors/exceptions.
fast respawn in both cases